### ISSUE TITLE
If BACKUP is rebooted, it transitions to MASTER state at same time as receiving MASTER advert from peer

### ISSUE THREAD

**Issue by dseeley** [acassen/keepalived#951]
I have a pair of servers, both configured as BACKUP initially.  After some time, in a steady state scenario, where dev-a0 is MASTER and the dev-a1 is BACKUP, I  perform a reboot of dev-a1, and observe the following

```
+-----------+                   +-----------+
|  dev-a0   |                   |  dev-a1   |
+----|------+                   +----|------+
     |                               |     
     |                               |     
    ...                             ...
     |                               |     
   MASTER                          BACKUP  
     |                               |     
     |                               |     
     |                               |     
     |                            <reboot>
     |   Master advert               |     
     |-----------------------X       |     
     |                             FAULT   
     |   Master advert               |     
     |-----------------------X       |     
     |                             BACKUP  
     |   Master advert               |     
     |-----------------------X       |     
     |                             MASTER  
     |   Master advert               |     
     |------------------------------>|     
     |                               |     
     |                               |     
     |                             BACKUP  
     |                               |     
```
This behaviour is a problem for me because I am using this in AWS, where the spurious MASTER transition on dev-a1 causes the EC2 secondary IP to be removed from dev-a0 and assigned to dev-a1, whereupon it is immediately removed when dev-a1 transitions back to BACKUP after receiving the advert; thus neither machine has the EC2 secondary IP.

I attach the logs of dev-a1.  The dev-a0 logs are uninteresting - nothing happens, it just stays in the MASTER state, constantly sending out master adverts.  In the dev-a1 keepalived and tcpdump logs, it appears that the vrrp advert from dev-a0 only starts being accepted by dev-a1 at the expiry of the advertisement timeout.  The Master advert arrives at exactly the same time as the advert expiry (no matter what the advert_int is set to - I tried several values).  

It appears that the vrrp port is only opened after the expiry of the first advert interval.  tcpdump shows dev-a1 accepting vrrp adverts only after this delay.  This problem only manifests after a reboot; VRRP adverts continue to be received by the host after stopping keepalived (is some module being loaded that stays resident after keepalived stops?).

Even if this is fixed (assuming it's actually a bug), what would also be a great enhancement, would be to have an option for notifies to be sent to all vrrp_instances on any state change (not only when their own state changes).  I think this would also help issue #950.

(Running version 2.0.5 on Ubuntu 18.04)

---
[backup_config.txt](https://github.com/acassen/keepalived/files/2183261/backup_config.txt)
[master_config.txt](https://github.com/acassen/keepalived/files/2183275/master_config.txt)
[backup-logs.txt](https://github.com/acassen/keepalived/files/2183256/backup-logs.txt)

**Comment by pqarmitage** [comment]
Unfortunately I have no experience of AWS, but what is happening here appears to be some peculiarity of AWS rather than keepalived, since it doesn't happen in "standard" environments.

From you description above, it appears that AWS isn't allowing packets for protocol 112 (VRRP) to be received until a packet has been sent using that protocol. As you say, it could be that a module is loaded (although I can't think of any module that is necessary), or it might be that the protocol isn't permitted through the firewall until a packet has been sent using that protocol (e.g. something like iptables --ctstate ESTABLISHED). This latter seems more likely to me.

Can you list the modules installed in the server before keepalived is loaded, and again after keepalived has loaded. Can you also list the firewall and conntrack settings before keepalived is loaded, and again afterwards.

Just as a thought for a possible workaround, it might be possible to add an option for keepalived to send a priority 0 advert when it first transitions to backup, and see if that then enables keepalived to receive adverts.

BTW, is there a reason you are running both vrrp instances with the same priority. It might be safer to set a higher priority on the one you prefer to be master.

**Comment by pqarmitage** [comment]
@dseeley I'm not entirely clear what you mean by `have an option for notifies to be sent to all vrrp_instances on any state change`. The way I see it, notifies are sent by running notify scripts, or sending a message to a notify fifo. Since you are using a notify_fifo, whatever is reading that fifo will receive all notifications for all vrrp instances. If someone is using a notify script, they can configure the same script for all notify events, and just pass a parameter that specifies the vrrp instance; the script can then handle the events as it chooses. Or am I misunderstanding what is being suggested?

**Comment by dseeley-sky** [comment]
Hi @pqarmitage - thanks for getting back so quickly.  Indeed, it looks like you're correct - the AWS security group was not by default open inbound for VRRP protocol - it appeared to be working, because the BACKUP, on transitioning to MASTER, sent out an advert, which opened up the firewall inbound, whereupon it immediately received the master advert.  Thanks!

They're both running with the same priority as they're provisioned automatically, and we really don't care which is master, as long as one of them is!

**Comment by dseeley** [comment]
Hi @pqarmitage - thanks for getting back so quickly. Indeed, it looks like you're correct - the AWS security group was not by default open inbound for VRRP protocol - it appeared to be working, because the BACKUP, on transitioning to MASTER, sent out an advert, which opened up the firewall inbound, whereupon it immediately received the master advert. Thanks!

They're both running with the same priority as they're provisioned automatically, and we really don't care which is master, as long as one of them is!

**Comment by dseeley** [comment]
Re the `have an option for notifies to be sent to all vrrp_instances on any state change`:
As I understand it, notify is triggered by a state change.  If server A changes state, it is told about it.  I was suggesting that the other servers in the group (B, C etc) could also be notified of their current state so as to remove any ambiguity. On reflection though, I'm not sure this would be very reliable, given the reason for the state change might be that the network may not be reliable and thus likewise, such a notification.

**Comment by pqarmitage** [comment]
The only protocol that is operating between the keepalived instances is the VRRP protocol, so we are limited to what information that can be derived from that. The possibilities that could be notified that currently aren't that I can think of are:

1. A master seeing an advert from another router that is a lower priority advert (that would have allowed a work around for the problem you were experiencing, but what we want to encourage is proper solutions such as you have now found!).
2. A change in source address or priority of an advert

I'm not sure how useful these would be, but if there is strong demand for them they could be considered.
